### PR TITLE
fix(simulation): the policy surface is discoverable from the schema and from a wrong first guess

### PR DESCRIPTION
**Pull request by cagataycali**
## What
`run_policy(policy="lerobot/act", task="wave")` was refused with a 22-name `Valid:` list including `observer` / `policy_object` and no hint. The refusal now names the nearest valid keys and lists only tool-reachable ones; `policy_provider`'s schema entry names every registered provider and says the model goes in `policy_config`; `instruction` is described.

## Why
Reachability is judged per top-level union member, not by substring over the whole annotation. `stop_when` is `dict[str, Any] | Callable[[SimEngine], bool] | None`: the schema documents its dict predicate DSL and a JSON dict reaches it, so matching the whole annotation would hide a published knob and misdirect the new hint.

refusal before and after

## Tests
`tests/simulation/mujoco/test_policy_surface_is_discoverable.py`, 27 cells, 5 failing pre-fix. Includes a grader that every published parameter of all 77 actions stays offered by its refusal - the direction `test_tool_spec_names_every_agent_knob.py` does not cover - and both schemas checked against the shipped provider registry. `tests/simulation/mujoco` 4,606 pass; ruff + mypy clean on 2,116 files.

## Review rounds

- **Round 1** (84c5a4ff1): `Test and Lint` was red on this head by one cell, `tests/registry/test_advertised_providers_are_registered.py::test_every_provider_schema_advertises_examples_in_a_readable_form`. That pre-existing grader read a schema's provider names only from an `e.g. a, b, c` clause, and this branch rewrites both schemas to the exhaustive `one of a, b, c` form, so it reported both as stating no list at all. The grader now reads both clauses (and stops at an opening bracket, so `(default groot)` is not read as a name), which grades every name in the exhaustive list for registration and self-resolution as it did a sample's: 28 parametrised cells where the two samples gave 8. One test file changed, no production change; 115 cells across both graders pass.
- **Round 2** (0d6ab715): `Test and Lint` was red by one cell of this branch's own grader, `TestTheSchemaNamesTheProviders::test_mujoco_schema_lists_every_registered_provider`: `providers registered but not in the schema description: ['custom_test', 'kwarg_test', 'preflight_*', ...]`. The cell compared the schema literal against `list_providers()`, which also reports every runtime `register_policy()` call - and in a whole-suite run those are the throwaway providers sibling tests register. The schema description is a literal written from `policies.json`, so both cells now grade against `list_policy_providers()` (the shipped registry) and the class docstring says why. Measured: pre-fix the cell passes alone and fails with one runtime registration present; post-fix it passes in both. Test-only, +16/-4. The same push absorbs `main` (clean merge, 0 conflict lines) so the branch carries #3743's `exited` field and clears `Detect an untested overlap with the base branch`; note `main` itself is red at `0583438` by two cells in `test_stop_policy_resolves_the_only_rollout.py` until #3798 lands, so this head's `Test and Lint` will report that failure, not this branch's, until it re-absorbs `main` after #3798.

### CODE REVIEW

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Makes the policy surface discoverable two ways: (1) an unknown-parameter refusal now carries a nearest-key hint (`policy` -> `policy_provider, policy_config`) and its `Valid:` list names only parameters a JSON tool call can actually carry, dropping Python-only callbacks and live objects (`observer`, `on_frame`, `policy_object`) via `_tool_call_can_carry`, which judges reachability per top-level union member so `stop_when` (`dict | Callable | None`) stays listed; (2) the `policy_provider` schema descriptions in both `tool_spec.json` and `hardware_robot.py` now enumerate every registered provider and point at `policy_config` for the model itself, and `instruction` gains a description. Verified against the head: the 14 hardcoded provider names match `registry/policies.json` exactly and sync is test-enforced in both schemas; the stated defaults match the actual signatures (sim `run_policy` defaults to `mock`, hardware to `groot`); `_tool_call_can_carry` fails open (keeps the key) for unannotated or unrecognised annotations, so the narrowed list cannot hide a usable key, and the new meta-test grades all 77 actions to pin that. No schema parameter names, defaults, or wire shapes change - description strings and refusal text only.

## What's good

- Both directions of the schema/refusal contract are graded: the existing `test_tool_spec_names_every_agent_knob.py` (no refusal advertises what the schema hides) plus the new all-actions sweep (no refusal hides what the schema publishes).
- The union-member reachability rule is documented with its motivating counterexample (`stop_when`) and pinned by parametrised tests, including the subscript-internal `|` case and the `PolicyConfig` word-boundary case.
- Changelog fragment included; ASCII-clean per project rules; `test_no_host_paths.py` passes on the head.

## Verification suggestions

- `python -m pytest tests/simulation/mujoco/test_policy_surface_is_discoverable.py tests/simulation/mujoco/test_tool_spec_names_every_agent_knob.py -q` - 33 tests, covers both grader directions (verified passing on `0f2cae7f`).
- Spot-check the live refusal: `Simulation` + `_dispatch_action("run_policy", {"policy": "lerobot/act"})` should answer with `Did you mean: policy_provider, policy_config?` and a `Valid:` list without `observer`/`policy_object`.

**Review by yinsong1986 — COMMENTED** [review]
## Summary

Makes the policy surface discoverable from two directions: an unknown-parameter refusal now carries a nearest-key hint (`policy` -> `policy_provider, policy_config`) and narrows its `Valid:` list to parameters a JSON tool call can actually carry, judged per top-level union member by `_tool_call_can_carry` so `stop_when` (`dict | Callable | None`) stays listed while pure callbacks and live objects (`observer`, `on_frame`, `policy_object`) drop out; the `policy_provider` schema descriptions in `tool_spec.json` and `hardware_robot.py` now enumerate the registered providers and point at `policy_config` for the model itself, and `instruction` gains a description. Production changes are description strings and refusal text only - no parameter names, defaults, or wire shapes move, and `_tool_call_can_carry` fails open (keeps the key) for unannotated or unrecognised annotations, with an all-actions sweep pinning that no published knob is hidden. The Round-1 grader fix in `tests/registry/test_advertised_providers_are_registered.py` correctly extends the clause reader to `one of ...` and stops at brackets.

## What's good

- Both directions of the schema/refusal contract are now graded (existing knob test plus the new 77-action sweep).
- The union-member reachability rule is documented with its motivating counterexample (`stop_when`) and pinned by parametrised tests, including the subscript-internal `|` and `PolicyConfig` word-boundary cases.
- Changelog fragment present, ASCII-clean, no host paths.

## Verification suggestions

- Reproduce the full-suite order dependence locally: `python -m pytest tests/policies/test_factory.py tests/simulation/mujoco/test_policy_surface_is_discoverable.py -q` (the schema-completeness cells fail when factory tests run first; they pass in isolation).
- The `Detect an untested overlap with the base branch` check asks for a merge of `main` (both `hardware_robot.py` and `simulation.py` moved on `main` since divergence); re-running the suite on the merged tree is worth doing before the next push.

File: tests/simulation/mujoco/test_policy_surface_is_discoverable.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    def test_a_union_inside_a_subscript_does_not_keep_a_callable(self):
+        """The ``|`` between a callback's own argument types is not an alternative."""
+        annotation = "Callable[[Started | Step | Ended], None] | None"
+        assert _tool_call_can_carry(self._param(annotation)) is False
+
+
+class TestTheSchemaNamesTheProviders:
+    def test_mujoco_schema_lists_every_registered_provider(self):
+        desc = _TOOL_SPEC_SCHEMA["properties"]["policy_provider"]["description"]
+        missing = [name for name in list_providers() if name not in desc]
```
Comment (yinsong1986) [inline tests/simulation/mujoco/test_policy_surface_is_discoverable.py]: [MUST FIX] This assertion grades the static schema description against the *live* provider registry, and the registry is a mutable global that other tests in the suite write into. In the full-suite run at this head (`Test and Lint`, job 104866699572) this cell fails with:

```
AssertionError: providers registered but not in the schema description: ['custom_test', 'kwarg_test', 'preflight_accept', 'preflight_agreement_probe', 'preflight_capture', 'preflight_overrides_probe', 'preflight_reject', 'runtime_only_provider', 'safe_custom', 'sc', 'strict_test', 'tolerant_test']
```

Every one of those names is a test-only provider registered by `tests/policies/test_factory.py` via `register_policy(...)` earlier in the same session, so the test is order-dependent: green in isolation, red in CI. `test_hardware_schema_lists_every_registered_provider` (line 162) has the same shape and only escapes because `-x` stops the run first. This blocks merge concretely - the required `Test and Lint` check is red at `84c5a4ff` because of this cell, and since the suite runs under `-x`, 61.6% of the suite behind it never executed, so the branch's test state is unknown past this point.

Fix: derive the expected provider set from the static source the schema is meant to mirror - `strands_robots/registry/policies.json` (the same module `tests/registry/test_advertised_providers_are_registered.py` anchors on via `resolve_policy`) - instead of live `list_providers()`; or filter the live list down to the built-in registry snapshot before comparing. Apply the same change to both schema-completeness cells.

Comment (cagataycali) [reply tests/simulation/mujoco/test_policy_surface_is_discoverable.py]: Answered on-branch by 0d6ab71, in the shape you asked for: both schema-completeness cells now derive their population from `list_policy_providers()` (the `policies.json` snapshot) instead of live `list_providers()`, and the live import is gone from the module.

Re-measured on the current head in one session in your order - `tests/policies/test_factory.py` first, so the throwaway providers really are registered:

| tree | result |
|---|---|
| head a4c624b | 56 passed |
| same session, `list_providers()` restored | 2 failed, 54 passed |

The control fails on both cells, including `test_hardware_schema_lists_every_registered_provider` - which `-x` had been hiding - naming `custom_test`, `kwarg_test`, `preflight_*`, `runtime_only_provider`, `safe_custom`, `sc`.

**Review by yinsong1986 — DISMISSED** [review]
## Summary

Makes the policy surface discoverable from both ends of a tool call: an unknown-parameter refusal in `Simulation._validate_and_build_kwargs` now carries a `close_match_hint` nearest-key fragment (so `policy=` is answered with `policy_provider, policy_config`) and narrows its `Valid:` list to parameters a JSON tool call can carry, judged per top-level union member by the new `_tool_call_can_carry` / `_union_members` helpers so `stop_when` (`dict | Callable | None`) stays listed while pure callbacks (`observer`, `on_frame`) and live `Policy` objects are dropped. On the schema side, both `tool_spec.json` and `HardwareRobot.tool_spec` rewrite `policy_provider` from a sampled `e.g. ...` clause to an exhaustive `one of ...` list, and `instruction` gains a description. I verified the 14 names in both literals exactly match the shipped `strands_robots/registry/policies.json`, the stated defaults match the actual signatures (sim `run_policy` defaults to `mock`, hardware to `groot`), and the helpers fail open (`True`) on anything they cannot judge, so no published knob can be hidden by a misjudged annotation - and the new `TestNoPublishedKnobIsHiddenByARefusal` grader sweeps all actions to pin that. The existing `_eg_names` grader was widened to read the new `one of` form, keeping both directions (advertised names are registered; registered names are advertised) enforced by CI, so the duplicated literals cannot silently drift from the registry. The prior review's concern about grading the schema against the mutable live registry is resolved on this head via `list_policy_providers()`. Latest `Test and Lint` on this head is green; note the branch is currently unmergeable (`mergeable_state: dirty`) against moved `main`, so a re-absorb of `main` is needed before merge, as the PR description already anticipates.

## What's good

- Both directions of the schema/refusal contract are pinned by graders (`TestNoPublishedKnobIsHiddenByARefusal` here, `test_tool_spec_names_every_agent_knob.py` for the converse), so the narrowed list cannot regress into hiding a published key.
- `_tool_call_can_carry` defaults to keeping a parameter whenever the annotation is empty, unparseable, or `None`-only - the safe direction for a refusal surface.
- Regression tests cover the reviewed fix (static-registry grading), the `PolicyConfig` false-positive, and the union-inside-subscript case, per AGENTS.md > Review Learnings (#86) > Testing Patterns.
- Changelog entry included; error strings are plain ASCII.

**Comment by cagataycali** [comment]
Rebased onto `main` (`15b0d6fee`); now MERGEABLE, squash auto-merge armed. The only conflict was the `policy_provider` schema description, which #3768 rewrote too. Both texts are graded, so the resolution keeps both facts:

| `policy_provider` description | scope run |
|---|---|
| this branch's text alone | 1 F - `test_provider_parameter_carries_the_same_facts` (`groot (default, needs policy_port)` gone) |
| main's text alone | 1 F - `test_hardware_schema_lists_every_registered_provider` (10 providers unlisted) |
| the merged text | 126 P |

Merged text: the exhaustive `one of ...` clause, then main's sentences on which providers dial a server and what each still needs.

Gate: ruff clean (2154 files), mypy `Success: no issues found in 2154 source files`, `tests/simulation` + `tests/registry` + hardware description/lifecycle 17292 P / 199 S / 0 F, whole-tree graders 5918 P / 49 S / 0 F.

**Review by yinsong1986 — APPROVED** [review]
## Summary

Makes the policy surface discoverable from both ends of a tool call. An unknown-parameter refusal in `Simulation._validate_and_build_kwargs` now appends a `close_match_hint` nearest-key fragment (so `policy=` is answered with `policy_provider, policy_config`) and narrows its `Valid:` list to parameters a JSON tool call can carry, judged per top-level union member by the new `_tool_call_can_carry` / `_union_members` helpers so `stop_when` (`dict | Callable | None`) stays listed while pure callbacks and live `Policy` objects (`observer`, `on_frame`, `policy_object`) drop out. Both `policy_provider` schema entries (sim `tool_spec.json` and `hardware_robot.py`) now list every registered provider exhaustively - I verified the 14-name list matches `policies.json` at this head, and the stated defaults (`mock` for sim, `groot` for hardware) match the actual signature defaults - and `instruction` gains a description that captures the `task=`/`prompt=` guess. The prior must-fix (grading the static schema against the live, test-polluted registry) is resolved: both completeness cells now grade against `list_policy_providers()`, with the reasoning documented in the class docstring. Drift is guarded in both directions - the new test fails if a registered provider is missing from the schema, and the updated `_eg_names` grader in `test_advertised_providers_are_registered.py` reads the `one of ...` clause name-by-name against the registry (stopping at an opening bracket so `(default mock ...)` is not read as a name).

## What's good

- The fallback contract in `_tool_call_can_carry` is safety-first: unannotated parameters and anything ambiguous stay listed, so the narrowed refusal can only over-offer, never hide a key - and `TestNoPublishedKnobIsHiddenByARefusal` sweeps all published parameters of every action to pin that invariant.
- `_union_members` splits only at bracket depth zero, so `Callable[[Started | Step | Ended], None] | None` is correctly judged as two members, with a regression test for exactly that case.
- Changelog fragment present, no emojis in user-facing strings, no host paths in tests, per AGENTS.md.

## Verification suggestions

The `Detect an untested overlap with the base branch` check is red because `main` moved again (`hardware_robot.py`, `simulation.py`, `tool_spec.json`) after the last merge at `15b0d6fe`; one more merge of `main` + push clears it and re-runs `Test and Lint` against the combined tree before merge.
